### PR TITLE
Allow fileName to be an Array of Function

### LINKED ISSUE #58: Allow files to be an array
So that you do things like this

```js
var json_data = io.readDataSync([__dirname, 'path', 'to', 'data.csv'])
```

### PR DESCRIPTION
**Pull request by aubergene**
Accept array or function for fileName. Work in progress

Do you think `src/utils/path.js` is the right approach?

Closes #58

### CODE REVIEW

**Comment by mhkeller** [comment]
Cool. Looks like `discernPath` still needs to be exported for the CI to check the tests. We should also probably include a test for each function that this could apply to in order to ensure coverage. Perhaps there's a better way to structure tests to allow for more variation like this without writing a bunch of new things.

**Comment by aubergene** [comment]
Ah yes. I didn't check in a build version. Yeah, I was unsure what is easiest with the tests, perhaps we could add a helper method into the tests which cycles over the different ways of constructing a path.

Adding an `.editorconfig` would be great

I got some whitespace diff issue for generating Yaml when running tests. I think it's because when I did npm install I got a different version (js-yaml 3.9.1). Perhaps add a lock file?

I also got an issue about `rollup-plugin-node-builtins` missing for build:browser, but I'm guessing it's still a work in progress

**Comment by aubergene** [comment]
Do you think this functionality and approach looks good though in general?

**Comment by mhkeller** [comment]
Ya I think so. I can't think of a simpler alternative. Maybe @ebrelsford has a thought that I'm not seeing?

**Comment by mhkeller** [comment]
what was the `rollup-plugin-node-builtins ` problem? Maybe open that up as an issue.

File: src/utils/path.js [hunk 1]
```diff
@@ -25,6 +25,6 @@ export function dirname (path) {
 }
 
 export function joinPath () {
-  var args = Array.prototype.slice.call(arguments)
+  var args = (arguments.length === 1 ? arguments[0] : Array.apply(null, arguments))
```
Comment (mhkeller) [inline src/utils/path.js]: This doesn't join paths that are strings: https://codepen.io/anon/pen/gxNGjv

**Review by mhkeller — CHANGES_REQUESTED** [review]

**Comment by mhkeller** [comment]
@aubergene FYI the yaml and rollup plugin issues should be fixed. There's also now an .editorconfig and package-lock.json. I think the next step for this is:

1. Fix `joinPath` per above
2. Add tests for all existing functions that take a filePath

**Comment by aubergene** [comment]
I think this is good now. The tests currently use a mix of strings and arrays but do the joining themselves, so we could just remove `path.join`.

**Comment by mhkeller** [comment]
Looks good overall. If you do the below, I'll update the docs if that sounds like a deal. It'd be nice to include this in 2.1.0. 

What do you mean by we could remove `path.join`? Do you mean test other functions by changing how they load in data. It would be cleaner to add two tests to each function so there isn't overlap.

I think something like `getPath` could be a better name. The `discern` functions are all about choosing a function based on evidence in the path. This function does something to a path.

1. Needs to be added to `readData`, `readDataSync` and `readdir`. The latter is used by both `readdirFilter` and `readdirFilterSync` so no need to add them there.
2. Needs tests using the functions affected. For example, something like this for `readDataSync` if you inserted this at the top on line 27:

```js
var testDataArr = function (morePath) {
  return [__dirname, 'data'].concat(morePath)
}
```

and then these tests at line 580

```js
    describe('json', function () {
      it('should match expected json with array path', function () {
        var json = io.readDataSync(testDataArr(['json', 'basic.json']))
        assertBasicValid(json)
      })
    })

    describe('json', function () {
      it('should match expected json with function path', function () {
        var json = io.readDataSync(function () {
          return testDataPath('json/basic.json')
        })
        assertBasicValid(json)
      })
    })
```

**Comment by aubergene** [comment]
Yeah, the renaming is good. I started on it, and then started thinking more about problems when allowing a function. We have code paths where it would get called more than once. For instance in `appendData` which appears to read the data, append the new data and then write it out again. It would be tricky to structure this so that it would only call `getPath` once and also so that `readData` and `writeData` would do same too. I think either we drop the idea of it being a function and just join it. Hmm starting to question whether this is a good idea at all now

**Comment by mhkeller** [comment]
Any updated thoughts on whether we should pursue this? Why would it have to be called more than once? After it gets called the first time it becomes a string and then `getPath` would simply return it and not join it again, no?

**Pull request closed without merging**